homeassistant/components/weather/__init__.py

```diff
@@ -114,37 +220,182 @@ class WeatherEntity(Entity):
+        """Post initialisation processing."""
+        super().__init_subclass__(**kwargs)
+        _reported = False
+        if any(
+            method in cls.__dict__
+            for method in (
+                "_attr_temperature",
+                "temperature",
+                "_attr_temperature_unit",
+                "temperature_unit",
+                "_attr_pressure",
+                "pressure",
+                "_attr_pressure_unit",
+                "pressure_unit",
+                "_attr_wind_speed",
+                "wind_speed",
+                "_attr_wind_speed_unit",
+                "wind_speed_unit",
+                "_attr_visibility",
+                "visibility",
+                "_attr_visibility_unit",
+                "visibility_unit",
+                "_attr_precipitation_unit",
+                "precipitation_unit",
+            )
+        ):
+            if _reported is False:
+                module = inspect.getmodule(cls)
+                _reported = True
+                if (
+                    module
+                    and module.__file__
+                    and "custom_components" in module.__file__
+                ):
+                    report_issue = "report it to the custom component author."
+                else:
+                    report_issue = (
+                        "create a bug report at "
+                        "https://github.com/home-assistant/core/issues?q=is%3Aopen+is%3Aissue"
+                    )
+                _LOGGER.warning(
+                    "%s::%s is overriding deprecated methods on an instance of "
+                    "WeatherEntity, this is not valid and will be unsupported "
+                    "from Home Assistant 2022.10. Please %s",
+                    cls.__module__,
+                    cls.__name__,
+                    report_issue,
+                )
+
+    async def async_internal_added_to_hass(self) -> None:
+        """Call when the sensor entity is added to hass."""
+        await super().async_internal_added_to_hass()
+        if not self.registry_entry:
+            return
+        self.async_registry_entry_updated()
 
     @property
     def temperature(self) -> float | None:
-        """Return the platform temperature in native units (i.e. not converted)."""
+        """Return the temperature for backward compatibility.
+
+        Should not be set by integrations.
+        """
         return self._attr_temperature
 
     @property
-    def temperature_unit(self) -> str:
+    def native_temperature(self) -> float | None:
+        """Return the temperature in native units."""
+        return self._attr_native_temperature
+
+    @property
+    def native_temperature_unit(self) -> str | None:
         """Return the native unit of measurement for temperature."""
-        return self._attr_temperature_unit
+        return self._attr_native_temperature_unit
+
+    @property
+    def temperature_unit(self) -> str:
```

Comment:
Do like this for backwards compatibility:
```suggestion
        """Return the temperature unit for backward compatibility.
        Should not be set by integrations.
        """
        return self._attr_temperature_unit

    @property
    def _temperature_unit(self) -> str:
```